rj_gameplay/reg_pass_costing/cost_heuristic.py

```diff
@@ -0,0 +1,157 @@
+import stp.rc as rc
+import rj_gameplay.gameplay_node as gameplay_node
+import rclpy
+import numpy as np
+
+MAX_PASS_DIST = 5
+MAX_GOAL_DIST = 7.5
+MAX_AVG_OPP_DIST = 6
+MAX_NEAREST_OPP_DIST = 2.5
+MAX_OPP_BETWEEN_ROBOT_GOAL = 5
+MAX_OPP_BETWEEN_ROBOT_BALL = 1
+MIN_GOAL_DIST_BONUS = 1
+MAX_GOAL_DIST_BONUS = 1.75
+
+GOAL_DIST_COST = 150
+AVG_OPP_DIST_COST = 75
+NEAREST_OPP_DIST_COST = 300
+OPP_BETWEEN_ROBOT_GOAL_COST = 325
+OPP_BETWEEN_ROBOT_BALL_COST = 650
+
+MAX_COST = 1000
+
+def cost_heuristic(world_state: rc.WorldState, robot: rc.Robot) -> float:
+
+    field = world_state.field
+
+    our_robots = world_state.our_robots
+    their_robots = world_state.their_robots
+
+    robot_with_ball = our_robots[0]
+    for bot in our_robots:
+        if bot.has_ball_sense:
+            robot_with_ball = bot
+            if robot_with_ball == robot:
+                return -1
+
+    # return 1 if robot is too far away from robot_with_ball
+    if (distance_from_ball(robot, robot_with_ball) > MAX_PASS_DIST):
+        return 1 
+
+    running_cost = 0
+    running_cost += distance_from_goal(robot, robot_with_ball, field)
+    running_cost += average_opponent_distance(robot, their_robots)
+    running_cost += nearest_opponent_distance(robot, their_robots)
+    running_cost += opponents_between_robot_goal(robot, their_robots, field)
+    running_cost += opponents_between_robot_ball(robot, their_robots, robot_with_ball)
```

Comment:
I don't know how much you've tested this, but it might be good to have some float coefficients by each of the variables. So it would look something like this:
'running_cost += goal_dist_coeff * distance_from_goal' and so forth. This would make it easier to tune in the future, and we can maybe even let people specify the weights of each if maybe someone wants to prioritize a robot super close to goal or something. For now, you can just define them all to 1.

Reply:
Those coefficients seem like they would essentially serve the same purpose as the cost values that I have defined at the beginning of my function, but I do the multiplication within each of the individual cost functions instead. I figured that if anyone wanted to tune the function in the future, they could just adjust those cost values.